framework/infrastructure/gcp/cloud_run.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2025 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import dataclasses
+import logging
+from typing import Any, Dict, Final
+
+from googleapiclient import discovery
+
+from framework.infrastructure import gcp
+
+logger = logging.getLogger(__name__)
+
+# Type aliases
+GcpResource = gcp.compute.ComputeV1.GcpResource
+
+DEFAULT_TEST_PORT: Final[int] = 8080
+DISCOVERY_URI: Final[str] = "https://run.googleapis.com/$discovery/rest?"
+
+
+@dataclasses.dataclass(frozen=True)
+class CloudRun:
+    service_name: str
+    url: str
+
+    @classmethod
+    def from_response(cls, name: str, d: Dict[str, Any]) -> "CloudRun":
+        return cls(
+            service_name=name,
+            url=d["urls"],
+        )
+
+
+class CloudRunApiManager(
+    gcp.api.GcpStandardCloudApiResource, metaclass=abc.ABCMeta
+):
+    project: str
+    region: str
+    _parent: str
+    service: discovery.Resource
+    api_manager: gcp.api.GcpApiManager
```

Comment:
all these are unused outside of the __init__ method, please convert to local variables